### PR TITLE
Support model download from public blobstore endpoints

### PR DESCRIPTION
**Pull request by novaturient95**

### CODE REVIEW

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml [hunk 1]
```diff
@@ -1,11 +1,8 @@
 path:                                                                 #Required
-  # local: "./local-path-to-model-files"                              #Required  Either remote or local should be passed. if both are given, local is given preference  
-  remote:                                                             #Required. Either remote or local should be passed. if both are given, local is given preference  
-    type: git                                                         #Required. Download Type. Can be git, http or ftp
-    uri: https://huggingface.co/bert-base-uncased                     #Required. Download URI
-    commit_hash: 5546055f03398095e385d7dc625e636cc8910bf2             #Optional
+  type: git                                                           #Required
```
Comment (wayliums) [inline inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:2]: i thought you guys said EV2 doesn't support this?

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml [hunk 2]
```diff
@@ -1,11 +1,8 @@
 path:                                                                 #Required
-  # local: "./local-path-to-model-files"                              #Required  Either remote or local should be passed. if both are given, local is given preference  
-  remote:                                                             #Required. Either remote or local should be passed. if both are given, local is given preference  
-    type: git                                                         #Required. Download Type. Can be git, http or ftp
-    uri: https://huggingface.co/bert-base-uncased                     #Required. Download URI
-    commit_hash: 5546055f03398095e385d7dc625e636cc8910bf2             #Optional
```
[the remark this answers is not in the document]
Comment (wayliums) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:6]: and i think you probably want to capture the hash in the asset itself too, it makes it easier for you to do linear tracking

[the remark this answers is not in the document]
Comment (novaturient95) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:6]: I was thinking it will be great to capture commit hash at our end instead of publisher sharing that in config. Unless we have a requirement that we want to register model with changes upto a particular commit, and not the latest one all the time. Please let me know your thoughts as well. 

cc: @skasturi @praveenmathamsetty

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml (hunk 1 above)
Comment (novaturient95) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:2]: We can still support git for smaller models.

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (skasturi) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:6]: As an infrastructure, I believe adding some kind of hash is necessary when 3P developers/customers try to onboard their models and as @wayliums mentioned, tracking.

That said, `commit_hash` is relevant only in the context of `git`. 

I would suggest having multiple options: 

1. checksum (different algos) - http based downloads
2. commit_hash (git)

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml (hunk 1 above)
Comment (skasturi) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:2]: Is `git` installed available on Ev2 agents?

File: scripts/install_azcopy.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh -x
+pwd
```
Comment (skasturi) [inline scripts/install_azcopy.sh:2]: Do we need to have this script in the public repo?

Could we please move this to `azureml-asset` instead?

File: scripts/azureml-assets/azureml/assets/config.py [hunk 4]
```diff
@@ -260,12 +260,84 @@ class ModelTaskName(Enum):
...
+    """Enum for path types supported for model publishing."""
 
-    FTP = 'ftp'
-    GIT = 'git'
-    HTTP = 'http'
+    LOCAL = "local"  # Path to model files present locally.
+    GIT = "git"      # Model hosted on a public GIT repo and can be cloned by GIT LFS.
+    FTP = "ftp"      # <UNSUPPORTED> Model files hosted on a FTP endpoint.
+    HTTP = "http"    # <UNSUPPORTED> Model files hosted on a HTTP endpoint.
+    AZURE = "azure"  # Model files hosted on an AZURE blobstore with public read access.
```
Comment (skasturi) [inline scripts/azureml-assets/azureml/assets/config.py:270]: >AZURE = "azure"  # Model files hosted on an AZURE blobstore with public read access. [](http://example.com/codeflow?start=4&length=84)

Wondering if we want to rename this to `azureblob` as we will add support for more data stores in the future.

Comment (skasturi) [reply scripts/azureml-assets/azureml/assets/config.py:270]: https://azure.microsoft.com/en-in/products/

image

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml [hunk 5]
```diff
@@ -1,11 +1,8 @@
...
+  type: git                                                           #Required
+  uri: https://huggingface.co/bert-base-uncased                       #Required
+  branch: main                                                        #Required
 publish:                                                              #Required
   type: custom_model                                                  #Required
   #flavor: hftransformers                                             #Required if type==mlflow_model 
   #task_name: huggingface_task_name                                   #Required if flavor==hftransformers
```
Comment (skasturi) [inline inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:8]: Could we remove this?

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py [hunk 6]
```diff
@@ -49,33 +42,40 @@ def _run(self, cmd, cwd: Path = "./") -> int:
...
-                # TODO: Error handling incase of incorrect commit hash.
-                logger.log_warning("Commit hash doesn't exist. Using model from latest HEAD.")
-        git_path = os.path.join(self.model_dir, '.git')
+        git_path = model_dir / ".git"
         shutil.rmtree(git_path, onerror=_onerror)
         return True
 
-    def download_model(self) -> bool:
+    def _download_azure_artifacts(model_url, model_dir) -> bool:
+        """Download model files from blobstore."""
```
Comment (skasturi) [inline scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:61]: (nit) Could you please add docs for the params?

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py [hunk 7]
```diff
@@ -49,33 +42,40 @@ def _run(self, cmd, cwd: Path = "./") -> int:
             logger.print(f"Successfully executed! Output: \n{result.stdout}")
         return result.returncode
 
-    def _download_git_model(self) -> bool:
+    def _download_git_model(model_url, model_dir) -> bool:
```
Comment (skasturi) [inline scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:45]: (nit) Please add docs for updated params

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py [hunk 8]
```diff
@@ -49,33 +42,40 @@ def _run(self, cmd, cwd: Path = "./") -> int:
...
 
-    def download_model(self) -> bool:
+    def _download_azure_artifacts(model_url, model_dir) -> bool:
+        """Download model files from blobstore."""
+        commands = f"""
+        azcopy login --service-principal --application-id $SP_CLIENT_ID --tenant-id $SP_TENANT_ID
+        azcopy cp --recursive=true {model_url} {model_dir}
+        """
+        result = ModelDownloadUtils._run(commands)
+        # TODO: Handle error case correctly, since azcopy exits with 0 exit code, even in case of error.
```
Comment (skasturi) [inline scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:67]: ># TODO: Handle error case correctly, since azcopy exits with 0 exit code, even in case of error. [](http://example.com/codeflow?start=8&length=96)

Could we please create a separate gh issue & link it here?

**Comment by skasturi** [comment]
> name: Log in to Azure and create resources

Are there any edits being done in this file?

---
Refers to: .github/actions/create-azure-resources/action.yaml:1 in 44e129b. [](commit_id = 44e129b9b24b17adddc8c118550ff56e1b316be6, deletion_comment = False)

File: scripts/azureml-assets/azureml/assets/__init__.py [hunk 9]
```diff
@@ -7,12 +7,12 @@
     Config,
     DEFAULT_ASSET_FILENAME,
     ModelConfig,
-    ModelDownloadType,
```
Comment (skasturi) [inline scripts/azureml-assets/azureml/assets/__init__.py:10]: Why are we deleting this? I remember @haritha-bendapudi facing an issue in the pipeline and having to send out another PR to include the imports. 

@haritha-bendapudi Could you please comment?

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (skasturi) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:6]: .. optional though.

**Review by skasturi — CHANGES_REQUESTED** [review]
:clock1:

File: scripts/azureml-assets/azureml/assets/config.py [hunk 10]
```diff
@@ -260,12 +260,84 @@ class ModelTaskName(Enum):
...
+        Args:
+            uri (str): Path to local model relative to asset.yaml.
+        """
+        super().__init__(PathType.LOCAL, uri=uri)
+
+
+class AzureBlobstoreAssetPath(AssetPath):
+    """Azure Blobstore asset path."""
+
+    BLOBSTORE_URI = "https://{}.blob.core.windows.net/{}/{}"
```
Comment (skasturi) [inline scripts/azureml-assets/azureml/assets/config.py:312]: > BLOBSTORE_URI = "https://{}.blob.core.windows.net/{}/{}" [](http://example.com/codeflow?start=3&length=57)

Could you please create an issue to support ABFS and WASB and add a TODO here? 

[https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-abfs-driver](https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-abfs-driver)

File: scripts/azureml-assets/azureml/assets/__init__.py (hunk 9 above)
Comment (haritha-bendapudi) [reply scripts/azureml-assets/azureml/assets/__init__.py:10]: Not in this pipeline @skasturi . In the ADO pipeline, they replace this __init__.py with another test __init__.py. Anything that is removed from __init__.py in this file has to be deleted/added in https://msdata.visualstudio.com/Vienna/_git/azureml-asset?path=/scripts/replace/__init__.py&version=GBmain&_a=contents

File: scripts/azureml-assets/azureml/assets/config.py (hunk 4 above)
Comment (novaturient95) [reply scripts/azureml-assets/azureml/assets/config.py:270]: Make sense. Modified to azureblob

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml (hunk 1 above)
Comment (novaturient95) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:2]: we can install git in the requirements if not already available. Similar to `az-copy`

File: scripts/azureml-assets/azureml/assets/config.py (hunk 10 above)
Comment (novaturient95) [reply scripts/azureml-assets/azureml/assets/config.py:312]: Done.

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml (hunk 5 above)
Comment (novaturient95) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:8]: Done

File: scripts/install_azcopy.sh (hunk 3 above)
Comment (novaturient95) [reply scripts/install_azcopy.sh:2]: Was seeing if this will get run as part of assets-test.yml in workflow. But it is getting skipped as label `safe to test` is not set. @elesel @wayliums can you please help to share how to add label `safe to test`. And also, if we can do model assets validation as part of publisher spec PR. In that case we will require `azcopy` installation if model is downloaded and registered to a test workspace.

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py (hunk 8 above)
Comment (novaturient95) [reply scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:67]: Created #283

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py (hunk 7 above)
Comment (novaturient95) [reply scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py]: Done

**Comment by novaturient95** [comment]
> Are there any edits being done in this file?

removed

**Comment by novaturient95** [comment]
> > name: Log in to Azure and create resources
> 
> Are there any edits being done in this file?
> 
> Refers to: .github/actions/create-azure-resources/action.yaml:1 in [44e129b](https://github.com/Azure/azureml-assets/commit/44e129b9b24b17adddc8c118550ff56e1b316be6). [](commit_id = [44e129b](https://github.com/Azure/azureml-assets/commit/44e129b9b24b17adddc8c118550ff56e1b316be6), deletion_comment = False)

removed

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py (hunk 6 above)
Comment (novaturient95) [reply scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py]: Done

**Review by wayliums — DISMISSED** [review]

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml (hunk 1 above)
Comment (novaturient95) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml]: Resolving this. Current models are to be downloaded from blobstore

File: inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (novaturient95) [reply inference/mlflow-huggingface/src/models/bert-base-uncased/model.yaml:6]: tracking separately.

**Review by quchuyuan — DISMISSED** [review]

**Review by quchuyuan — APPROVED** [review]

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py [hunk 11]
```diff
@@ -49,33 +42,64 @@ def _run(self, cmd, cwd: Path = "./") -> int:
...
+
+        :param model_url: Publicly readable blobstore URI of model files
+        :type model_url: str
+        :param model_dir: local directory to download model to
+        :type: Path
+        """
+        try:
+            SP_CLIENT_ID = os.environ["SP_CLIENT_ID"]
+            SP_TENANT_ID = os.environ["SP_TENANT_ID"]
+            login = f"azcopy login --service-principal --application-id {SP_CLIENT_ID} --tenant-id {SP_TENANT_ID}"
```
Comment (skasturi) [inline scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:72]: Does azcopy support streaming the data from the url to the target storage we are hosting in?

Comment (skasturi) [reply scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:72]: We might have to re-think after this PR, if we prefer to completely download the model to the agent & then upload vs stream to the target storage. 

Especially when we are going to support larger models.

Comment (skasturi) [reply scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:72]: Could you please create an issue if one doesn't already exist for this?

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py [hunk 12]
```diff
@@ -49,33 +42,64 @@ def _run(self, cmd, cwd: Path = "./") -> int:
...
+            result = ModelDownloadUtils._run(login)
+            if result:
+                logger.log_error("azcopy failed to login")
+                return False
+            download_cmd = f"azcopy cp --recursive=true {model_uri} {model_dir}"
```
Comment (skasturi) [inline scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:77]: Could you please update this command to output verbose logs?

**Review by skasturi — APPROVED** [review]

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py (hunk 12 above)
Comment (novaturient95) [reply scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:77]: `azcopy` has default log verbosity set to INFO.

image

File: scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py (hunk 11 above)
Comment (novaturient95) [reply scripts/azureml-assets/azureml/assets/release/model_publish_utils/utils.py:72]: Created issue: https://github.com/Azure/azureml-assets/issues/285

File: scripts/install_azcopy.sh (hunk 3 above)
Comment (novaturient95) [reply scripts/install_azcopy.sh]: Resolving this. Checking with Chuyuan if we need in this repo
